metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,338 @@
+import scala.meta.internal.parsing.Trees
+import scala.meta.pc.CancelToken
+import scala.meta.transversers.SimpleTraverser
+
+import org.eclipse.{lsp4j => l}
+
+class ExtractRenameMember(
+    buffers: Buffers,
+    trees: Trees
+) extends CodeAction {
+
+  override def contribute(params: l.CodeActionParams, token: CancelToken)(
+      implicit ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = {
+    val uri = params.getTextDocument.getUri
+    val path = uri.toAbsolutePath
+    val range = params.getRange
+
+    Future {
+      trees.get(path) match {
+        case Some(tree) =>
+          // Get the file name without extension
+          val fileName = uri.toAbsolutePath.filename.replaceAll("\\.scala$", "")
+
+          val definitions = membersDefinitions(tree)
+          val sldNames = sealedNames(tree)
+          val defnAtCursor =
+            definitions.find(_.name.pos.toLSP.overlapsWith(range))
+
+          def canRenameDefn(defn: Member): Boolean = {
+            val differentNames = defn.name.value != fileName
+            val notGetsExtendedByFileName =
+              !getExtendsNames(defn).contains(fileName)
+            differentNames && notGetsExtendedByFileName && defnAtCursor.exists(
+              _.equals(defn)
+            )
+          }
+
+          def canExtractDefn(defn: Member): Boolean = {
+            val differentNames = defn.name.value != fileName
+            val notExtendsSealedOrSealed = notSealed(defn, sldNames)
+            val companion = definitions.find(c => {
+              !c.equals(defn) && c.name.value.equals(defn.name.value)
+            })
+            val companionNotSealed =
+              companion.exists(notSealed(_, sldNames)) || companion.isEmpty
+
+            differentNames && notExtendsSealedOrSealed && companionNotSealed
+          }
+
+          definitions match {
+            case Nil => Nil
+            case head :: Nil if canRenameDefn(head) =>
+              definitions.map(renameFileAsMemberAction(uri, _))
+            case _ =>
+              defnAtCursor
+                .filter(defn => canExtractDefn(defn))
+                .map { classDefn =>
+                  extractClassAction(uri, classDefn)
+                }
+                .toList
+          }
+
+        case _ => Nil
+      }
+
+    }
+  }
+
+  private def membersDefinitions(tree: Tree): List[Member] = {
+    val nodes: ListBuffer[Member] = ListBuffer()
+
+    val traverser = new SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg =>
+          super.apply(p)
+        case c: Defn.Class => nodes += c
+        case t: Defn.Trait => nodes += t
+        case o: Defn.Object => nodes += o
+        case e: Defn.Enum => nodes += e
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+    traverser(tree)
+
+    nodes.toList
+  }
+
+  private def sealedNames(tree: Tree) = tree.collect {
```

Comment:
The issue here is that `collect` signature is:
```scala
def collect[T](fn: PartialFunction[Tree, T]): List[T]
```

By using the suggested version changing it using a PF we'd have a `List[Boolean]`, that in the `notSealed` function  would not work to get the `memberExtendsSealed`, because it uses the sealed nodes' `name.value` to compare it with the `inits` (extends and with) of the candidate `Member`
But the `isSealed` function would definitely work to avoid code repetition here and in the next part of the code